### DISCUSSION TITLE
TCA crash : Release build with dynamic frameworks app

### DISCUSSION THREAD

**Discussion by PierreCapo** [pointfreeco/swift-composable-architecture#1680] [General]
### Description

We are currently using TCA along with [Tuist](https://github.com/tuist) in our application. In particular, a bit a la isowords we have split our app in multiple frameworks. However, in release builds (debug works fine), TCA crashes when some reducers are in different frameworks. This is the graph of how we have set up the app to reproduce the bug.

![image](https://user-images.githubusercontent.com/26744253/202856140-ffcaef45-a0f8-400c-b7b8-5e1449ceaeaa.png)

Thanks in advance for the help 🙌

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

App shouldn't crash in Release mode

### Actual behavior

App crashes in release mode. See attached video:

https://user-images.githubusercontent.com/26744253/202856365-75cd0ba7-b3f4-48f8-b590-0c2d86b49b7c.mp4

### Steps to reproduce

For simplicity and clarity, I have created a minimal project to reproduce the bug : https://github.com/PierreCapo/tca-frameworks-bug.
You can follow the video attached to exactly reproduce the crash.

### The Composable Architecture version information

0.45.0

### Destination operating system

iOS 14 +

### Xcode version information

14.0

### Swift Compiler version information

```shell
swift-driver version: 1.62.8 Apple Swift version 5.7 (swiftlang-5.7.0.127.4 clang-1400.0.29.50)
Target: arm64-apple-macosx13.0
```

**Comment by mbrandonw** [comment]
Hi @PierreCapo, it is not clear that this is an issue with the library. The project works fine when putting all the files into a single target. So, I am going to move it to a discussion and maybe someone who has experience with Tuist can help out.

I would also suggest asking on [Tuist's discussions](https://github.com/tuist/tuist/discussions). It could be an issue in Tuist, Xcode or maybe even Swift.

**Reply by PierreCapo** [reply]
Fair enough, I'll try to set up the same project without Tuist so we can eliminate (or not) this hypothesis 👍

**Comment by hmhv** [comment]
I think TCA is not good with dynamic framework. i tried your test repo with staticFramework and it worked.

https://github.com/PierreCapo/tca-frameworks-bug/blob/main/src/FeatureSettings/Project.swift#L22

```diff
-            product: .framework,
+            product: .staticFramework,
```

**Reply by PierreCapo** [reply]
That is a super interesting. Indeed I think you are correct, this is something about static vs dynamic frameworks. It leads me to test different scenarios and this is what I get at the end

1. If TCA is a _static framework_ (how Tuist install it by default) inside a _dynamic framework_ feature there is a crash in Release mode ❌.
2. If TCA is a _static framework_ inside a _static framework_  feature it is working ✅ (@hmhv proposal)
3. If TCA is a _dynamic framework_ inside a _dynamic framework_  feature it is working ✅

For future people coming into this discussion, this is how to achieve point 3 with Tuist to make it compatible with TCA and dynamic framework features :

```Swift
// Tuist/Dependencies.swift

import ProjectDescription

let dependencies = Dependencies(
    swiftPackageManager: SwiftPackageManagerDependencies(
        [
            .remote(url: "https://github.com/pointfreeco/swift-composable-architecture", requirement: .exact("0.45.0")),
        ],
        // add this line to use TCA as a dynamic framework
        productTypes: ["ComposableArchitecture": .framework],
        baseSettings: Settings.settings(
            configurations: [
                .debug(name: "Debug"),
                .release(name: "Release"),
            ]
        )
    ),
    platforms: [.iOS]
)
```

Please note that finally I haven't tested yet the 3 aforementioned point without Tuist, so while I am sure they are valid with a project setup with Tuist, I am not sure the same bugs and issues appear in a "vanilla project", so that could potentially be an issue on Tuist side.

Thanks @hmhv  for the hint

**Reply by mbrandonw** [reply]
Hi @PierreCapo, thanks for sharing your findings. That's very interesting.

The issue of static vs dynamic frameworks did come up recently (here https://github.com/pointfreeco/swift-composable-architecture/discussions/1657), and apparently it is intended behavior for Xcode to strip symbols from static frameworks if it thinks they aren't being used. Maybe that is what is happening with your original project.

One way to prevent this stripping is to add the `-ObjC` flag (see this [comment](https://github.com/apple/swift/issues/62056#issuecomment-1311930729)). Can you try that in your original project to see if that also fixes it?

**Reply by PierreCapo** [reply]
I have tried with both the `-ObjC` flag and the `-all-load` ones from the discussion you have sent and I haven't been able to make it work sadly. Still, many thanks for the link, this was worth a try and I have learnt a lot of things reading it.
